File: Java/954.java
__________________________________________________________________________________________________
sample 3 ms submission
class Solution {
    public boolean canReorderDoubled(int[] A) {
        int[] countNum = new int[20001];
        for (int a : A) {
            countNum[a + 10000] += 1;
        }

        for (int i = 1; i <= 10000; i++) {
            if (countNum[10000 + i] > 0){
                if (countNum[10000 + i * 2] >= countNum[10000 + i]) {
                    countNum[10000 + i * 2] -= countNum[10000 + i];
                } else {
                    return false;
                }
            }

            if (countNum[10000 - i] > 0) {
                if (countNum[10000 - i * 2] >= countNum[10000 - i]) {
                    countNum[10000 - i * 2] -= countNum[10000 - i];
                } else {
                    return false;
                }
            }
        }

        return true;
    }
}
__________________________________________________________________________________________________
sample 40604 kb submission
class Solution {
    public boolean canReorderDoubled(int[] A) {
        // count[x] = the number of occurrences of x in A
        Map<Integer, Integer> count = new HashMap();
        for (int x: A)
            count.put(x, count.getOrDefault(x, 0) + 1);

        // B = A as Integer[], sorted by absolute value
        Integer[] B = new Integer[A.length];
        for (int i = 0; i < A.length; ++i)
            B[i] = A[i];
        Arrays.sort(B, Comparator.comparingInt(Math::abs));

        for (int x: B) {
            // If this can't be consumed, skip
            if (count.get(x) == 0) continue;
            // If this doesn't have a doubled partner, the answer is false
            if (count.getOrDefault(2*x, 0) <= 0) return false;

            // Write x, 2*x
            count.put(x, count.get(x) - 1);
            count.put(2*x, count.get(2*x) - 1);
        }

        // If we have written everything, the answer is true
        return true;
    }
}
__________________________________________________________________________________________________

File: Python3/954.py
__________________________________________________________________________________________________
sample 648 ms submission
class Solution:
    def canReorderDoubled(self, A: List[int]) -> bool:
        c = collections.Counter(A)
        for x in sorted(c, key=abs):
            if c[x] > c[2 * x]:
                return False
            c[2 * x] -= c[x]
        return True
__________________________________________________________________________________________________
sample 14876 kb submission
class Solution:
    def canReorderDoubled(self, A: List[int]) -> bool:
        pos = dict()
        neg = dict()
        zer = 0
        for a in A:
            if a > 0:
                if a in pos:
                    pos[a] = pos[a] + 1
                else:
                    pos[a] = 1
            elif a < 0:
                if a in neg:
                    neg[a] = neg[a] + 1
                else:
                    neg[a] = 1
            else:
                zer = zer + 1
        
        if zer % 2:
            return False
        
        k = list(pos.keys())
        k.sort()
        for i in k:
            if pos[i] > 0: 
                if (i * 2 in pos) and (pos[2 * i] >= pos[i]):
                    pos[2 * i] = pos[2 * i] - pos[i]
                else:
                    return False
        
        k = list(neg.keys())
        k.sort(reverse=True)
        for i in k:
            if neg[i] > 0: 
                if (i * 2 in neg) and (neg[2 * i] >= neg[i]):
                    neg[2 * i] = neg[2 * i] - neg[i]
                else:
                    return False        
        return True
__________________________________________________________________________________________________
